torchdp/test/virtual_step_test.py

```diff
@@ -0,0 +1,311 @@
+
+    def test_clipper_accumulation(self):
+        """
+        Calling optimizer.virtual_step() should accumulate clipped gradients to form
+        one large batch.
+        """
+        self.setUp_privacy_engine(self.DATA_SIZE)
+
+        for x, y in self.dl:
+            logits = self.model(x)
+            loss = self.criterion(logits, y)
+            loss.backward()
+            self.optimizer.virtual_step()
+
+        # the full data gradient accumulated in .grad
+        grad = torch.cat(
+            [p.grad.reshape(-1) for p in self.model.parameters() if p.requires_grad]
+        ) * (self.BATCH_SIZE / self.DATA_SIZE)
+
+        self.optimizer.step()
+
+        # .grad should contain the average gradient over the entire dataset
+        accumulated_grad = torch.cat(
+            [p.grad.reshape(-1) for p in self.model.parameters() if p.requires_grad]
+        )
+
+        # the accumulated gradients in .grad without any hooks
+        orig_grad = self.effective_batch_grad
+
+        self.assertTrue(
+            torch.allclose(accumulated_grad, orig_grad, atol=10e-5, rtol=10e-3)
+        )
+        self.assertTrue(torch.allclose(grad, orig_grad, atol=10e-5, rtol=10e-3))
+
+    def test_mixed_accumulation(self):
+        """
+        Calling loss.backward() multiple times aggregates all per-sample gradients in
+        .grad-sample. Then, calling optimizer.virtual_step() should clip all gradients
+        and aggregate them into one large batch.
+        """
+        self.setUp_privacy_engine(self.DATA_SIZE)
+
+        for idx, (x, y) in enumerate(self.dl):
+            logits = self.model(x)
+            loss = self.criterion(logits, y)
+            loss.backward()
+
+            # accumulate per-sample grads for two mini batches
+            # and then aggregate them in the clipper
+            if (idx + 1) % 2 == 0:
+                self.optimizer.virtual_step()
+
+        # the full data gradient accumulated in .grad
+        grad = torch.cat(
+            [p.grad.reshape(-1) for p in self.model.parameters() if p.requires_grad]
+        ) * (self.BATCH_SIZE / self.DATA_SIZE)
+
+        self.optimizer.step()
+
+        # .grad should contain the average gradient over the entire dataset
+        accumulated_grad = torch.cat(
+            [p.grad.reshape(-1) for p in self.model.parameters() if p.requires_grad]
+        )
+
+        # the accumulated gradients in .grad without any hooks
+        orig_grad = self.effective_batch_grad
+
+        self.assertTrue(
+            torch.allclose(accumulated_grad, orig_grad, atol=10e-5, rtol=10e-3)
+        )
+        self.assertTrue(torch.allclose(grad, orig_grad, atol=10e-5, rtol=10e-3))
+
+    def test_zero_grad(self):
+        """
+        Calling optimizer.zero_grad() should erase any accumulated per-sample gradients.
+        """
+        self.setUp_privacy_engine(2 * self.BATCH_SIZE)
+
+        double_grads_per_sample = []
+        double_grads = []
```

Comment:
The same comment I put for commenting `double_grads` :)

Reply:
as above